icechunk/src/session.rs

```diff
@@ -1576,19 +1676,82 @@ impl<'a> FlushProcess<'a> {
     async fn write_manifest_for_existing_node(
         &mut self,
         node: &NodeSnapshot,
-        splits: ManifestSplits,
+        manifests: Vec<ManifestRef>,
+        old_snapshot: &Snapshot,
     ) -> SessionResult<()> {
-        let asset_manager = Arc::clone(&self.asset_manager);
-        let updated_chunks = updated_node_chunks_iterator(
-            asset_manager.as_ref(),
-            self.change_set,
-            self.parent_id,
-            node.clone(),
-        )
-        .await
-        .map_ok(|(_path, chunk_info)| chunk_info);
+        #[allow(clippy::expect_used)]
+        let splits =
+            self.splits.get(&node.id).expect("splits should exist for this node.");
+        let mut refs =
+            HashMap::<ManifestExtents, Vec<ManifestRef>>::with_capacity(splits.len());
+
+        let on_disk_extents =
+            manifests.iter().map(|m| m.extents.clone()).collect::<Vec<_>>();
+
+        let modified_splits = self
+            .change_set
+            .modified_manifest_extents_iterator(&node.id, &node.path)
+            .collect::<HashSet<_>>();
+
+        // FIXME: there is an invariant here
+        // ``modified_splits`` (i.e. splits used in this session)
+        // must be a subset of ``splits`` (the splits set in the config)
+
+        // TODO: this should be try_fold with the refs HashMap as state
+        for extent in splits.iter() {
+            if modified_splits.contains(extent) {
+                // this split was modified in this session, rewrite it completely
+                self.write_manifest_for_updated_chunks(node, extent)
+                    .await?
+                    .map(|new_ref| refs.insert(extent.clone(), vec![new_ref]));
+            } else {
+                let on_disk_bbox = on_disk_extents
+                    .iter()
+                    .filter_map(|e| e.intersection(extent))
+                    .reduce(|a, b| a.union(&b));
+
+                // split was unmodified in this session. Let's look at the current manifests
+                // and see what we need to do with them
+                for old_ref in manifests.iter() {
```

Comment:
it would help to rename the `manifests` arguments, to something that makes it clear which manifests are these